### ISSUE TITLE
Cannot infer generic argument type from passed callback

### ISSUE THREAD

**Issue by kolodny** [Microsoft/TypeScript#31146]
**TypeScript Version:**  3.4.0-dev.201xxxxx

**Search Terms:**

infer, parameter, argument, callback, function

**Code**

```ts
function inferArguments<T>(callback: ((t: T) => void)) {
  return callback;
}

function noop(){}

const explicit = inferArguments(({a = noop}: {a: Function}) => {});
explicit({a: noop}); // OK!
explicit({a: false}); // Expected error - Got one!

const implicit = inferArguments(({a = noop}) => {});
implicit({a: noop}); // OK!
implicit({a: false}); // Expected error - No Error!
```

**Expected behavior:**
Both function calls with `({a: false})` should cause a type error

**Actual behavior:**
Only the function call with the explicit typing causes a type error

[**Playground Link**](https://www.typescriptlang.org/play/#src=function%20inferArguments%3CT%3E(callback%3A%20((t%3A%20T)%20%3D%3E%20void))%20%7B%0A%20%20return%20callback%3B%0A%7D%0A%0Afunction%20noop()%7B%7D%0A%0Aconst%20explicit%20%3D%20inferArguments((%7Ba%20%3D%20noop%7D%3A%20%7Ba%3A%20Function%7D)%20%3D%3E%20%7B%7D)%3B%0Aexplicit(%7Ba%3A%20noop%7D)%3B%20%2F%2F%20OK!%0Aexplicit(%7Ba%3A%20false%7D)%3B%20%2F%2F%20Expected%20error%20-%20Got%20one!%0A%0Aconst%20implicit%20%3D%20inferArguments((%7Ba%20%3D%20noop%7D)%20%3D%3E%20%7B%7D)%3B%0Aimplicit(%7Ba%3A%20noop%7D)%3B%20%2F%2F%20OK!%0Aimplicit(%7Ba%3A%20false%7D)%3B%20%2F%2F%20Expected%20error%20-%20No%20Error!)

**Related Issues:** #30975 Looks similar but seems different since there should be a clear way for inference to work in this case

Note that `Parameters` is correctly able to extract the correct types for such a construct as seen in this bit of code:

```ts
function noop() { }
function callback({ a = noop }) { }
let args: Parameters<typeof callback>;
args[0].a as Function
```

This appears to be an error with inline callback functions used in this fashion

**Comment by kolodny** [comment]
Another thing I noticed is that this works for functions declared on the same scope but fails for function expressions passed in to the HOF

```ts
const inferArguments = <T>(callback: T) => callback;

function noop(){}

function takesNoop({ a = noop}) {}
const OK = inferArguments(takesNoop);
OK({a: noop}); // OK!
OK({b: noop}); // Expected error - Got one!
OK({a: false}); // Expected error - Got one!

const BAD = inferArguments(function takesNoop({ a = noop}) {});
BAD({a: noop}); // OK!
BAD({b: noop}); // Expected error - No Error!
BAD({a: false}); // Expected error - No Error!
```

**Comment by weswigham** [comment]
Set `noImplicitAny` to `true` and you'll see the issue is really that we issue a `noImplicitAny` error on `implicit` (though it seems like we infer all the types for `inferArguments` just fine, so it's odd).

**Comment by adrianheine** [comment]
I have a maybe related issue:
```ts
function f(a) { // Would emit an error with noImplicitAny
  return () => a
}
const x = f(1) // I would have expected `x` to be `() => number`, but it is `() => any`
const y: string = x() // Passes
```

```ts
function f<T>(a: T) {
  return () => a
}
const x = f(1) // `() => number`
const y: string = x() // Fails
```

**Comment by RyanCavanaugh** [comment]
Looks like a bug to me - the default in the binding pattern should have made a candidate for `T`

**Comment by pradyuman** [comment]
Any update on this?

**Comment by pradyuman** [comment]
Any update on this? I'm running into issues with callback type inference as well. Happy to provide more details and/or help debug if it makes sense.

**Comment by pradyuman** [comment]
Any update on this? I'm running into an issue that is perhaps related. In the following function, `Res` never gets inferred. 

```
export const execGrpc = <Req, Res extends {}>(
  call: (
    request: Req,
    callback: (error: ServiceError | null, response: Res) => void
  ) => ClientUnaryCall,
  req: Req
): Promise<Either<ServiceError, Res>>
```

Happy to provide more details and/or help debug if it makes sense.

**Comment by RobertWHurst** [comment]
I've also run into this same issue. It would be nice to see it fixed.

**Comment by theonlypwner** [comment]
@pradyuman I've also run into the same issue with making my own wrapper for gRPC calls. I found that it is due to having multiple overloads for `call` (the auto-generated TypeScript declarations have multiple overloads for generated methods in the client), which prevents inference from working.

Here's a reproducible example (tested with TypeScript 3.7.2):
```ts
function f<T, U> (
  call: (
    request: U,
    innerCallback: (error: Error | null, value?: T) => void
  ) => void,
  request: U): Promise<T> {
  return new Promise((resolve, reject) => call(request, (error, response) => {
    if (error) {
      reject(error)
    } else {
      resolve(response)
    }
  }))
}

declare function p (request: number, call: (error: null, a: number) => void)
// declare function p (request: number, a2: boolean, call: (error: null, a: number) => void)

async function g () {
  const v = await f(p, 1)
  // the type of v is inferred as number
  // but if the other declaration above is uncommented, it is inferred as unknown
}
```

**Comment by theonlypwner** [comment]
Another common issue is having to manually specify a type for the Promise constructor.

```ts
const p = new Promise((resolve) => resolve(1))
```

`p` is of type `Promise<unknown>`, so in order to use it later, one has to include the type:
```ts
const p = new Promise<number>((resolve) => resolve(1))
```

**Comment by Aetet** [comment]
@RyanCavanaugh this affects event react typings for typescript:
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/30057#discussion_r232123748

**Comment by theonlypwner** [comment]
Even C# and Java can infer the types:

```cs
using System;

namespace ConsoleApp1
{
    static class Program
    {
        static void f<T>(Action<T> c)
        {
            c(default);
        }

        static void Main()
        {
            f((string s) => Console.WriteLine(s ?? "test")); // output: test
        }
    }
}
```
```java
import java.util.function.Consumer;

public class Main
{
    static <T> T f(Consumer<T> c) {
        c.accept(null);
        return null;
    }
    public static void main(String[] args) {
        String str = f((String s) -> System.out.println(s)); // output: null
        // Integer i = f((String s) -> System.out.println(s)); // does not compile
    }
}
```

**Comment by xpl** [comment]
Ran into the same issue today! Any plans to get it fixed?

**Comment by anilanar** [comment]
@Aetet 

> @RyanCavanaugh this affects event react typings for typescript:
> [DefinitelyTyped/DefinitelyTyped#30057 (comment)](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/30057#discussion_r232123748)
> 
> Because now any usage useCallback leads to silent any inference
> 
> Any update on this?

That's because react typings don't provide typings for >TS 3.0.

For example, we use the following `useCallback` in our codebase instead of using "official" react typings:

```
export const useCallback: <T extends (...args: never[]) => unknown>(
  callback: T,
  deps: Array<unknown>,
) => T;
```

**Comment by Aetet** [comment]
@anilanar есть более правильный путь через extends Function:
https://www.typescriptlang.org/v2/en/play?#code/GYVwdgxgLglg9mABCAzgUwMIEMA2OBGWEA1ogDwAqiaAHlGmACYqIAUAdJ1gE4DmKALkRg0ANzTcA2gF0AlIgC8APmRhiYOAHcwS1gChEiCLgJFiQigBoDiRmgAOgxAEFu3LAE8y4dVp3XZC0QAbxtuNCgQbiRjPEISAG49AF89PVBIWARkdGw4swAmSmo6BmY2TnYefiEsMA8ZeWVEOo9dWNMSC0tbBydXdy8fDW0lQMQqUMNwyOijE3jiJNT08Gh4JFRMBbMAZjI8ztJaeiYWADE1rLAegCVdYCFDxZ7iNA8UAH5a+sannZIITCESiSGAyzSq0yG0QvAiAGU4ABbNCsapCNFCMAgJH4CQ9fBCFBQbgwMC8JoqURwGCMeRTRAzUEtPisACMPQA5AVObIIXoAPQClqIGAsETibgE0XisQSPQQBDEoyKHLbfIkVhogmUoEpHqNJKC4VYGUterSsXmjwKpVQWyqrbPQparA6xQqUL6xCGtJCkVWmBI+w4GAQGD21qWlhBkNhiPW21gZVoR25AHEXau93NL3JA1yI3+igeeypq32bhwfA4NBIxAAWkQUAAFlh7VbeHA4Iw9HCoIiUawIHzjRNS+WY7wNOFGI3RcHQ+HI-U+wjkai6UlixOzZo4NxiGTeIgy9xgGhoGuBxvWGg+UA

**Comment by LXSMNSYC** [comment]
![image](https://user-images.githubusercontent.com/4783372/93770519-dd612000-fc4e-11ea-87ac-f8037a1f1928.png)

Might be related. This requires `T` to be defined explicitly when using `test`.

**Comment by theonlypwner** [comment]
In TypeScript, this is the current issue (affecting `new Promise`):
```ts
// the original example in the original post seems to be OK now, as there is an error for
// const implicit = inferArguments(({a = noop}) => {});
function f<T>(_: (t: T) => void): T { return null as any; }
const x = f((_: string) => {}); // y has type string, OK

// current issue in TypeScript 3.7 to 4.0
function f2<T>(_: (t: (t: T) => void) => void): T { return null as any; }
const y = f2(x => x("1")); // y has type unknown, BAD <------------------------------
const y2 = f2((x: (x: string) => void) => x("1")); // y2 has type string, OK
const y3 = f2<string>(x => x("1")); // y3 has type string, OK
```

Also, it actually turns out that C# can infer the type in similar cases:
```cs
using System;

void f<T>(Action<Action<T>> c) { }
void g(Action<string> f) { }

f((Action<string> f) => f("a")); // ok
// f(f => f("a")); // error
// f(g); // error

void f2<T>(Action<T> c) { }
static void g2(string s) { }

f2((string _) => {}); // ok
// f2(g2); // error
```

**Comment by victorafael26** [comment]
Ran into this exact pattern in a real codebase today (React lazy-loading helper wrapping `React.lazy` with a `pick` option to support named exports). Confirms this is still present as of TypeScript 5.9.3 (also reproduced identically on the new Go-based compiler, 7.0.2) — 6+ years after this was filed and confirmed as a bug.

Minimal repro (isolated, no React):

```ts
declare function lazyLoad<T>(
  factory: () => Promise<any>,
  opts: { pick: (mod: any) => { default: T } }
): T;

declare const NamedComponent: (props: { foo: string }) => any;

// Extracted + explicitly annotated pick: infers T correctly
const pickFn = (m: any): { default: typeof NamedComponent } => ({ default: m.NamedComponent });
const A = lazyLoad(() => Promise.resolve({}), { pick: pickFn });
const a: string = A; // correctly errors: (props) => any is not string

// Inline pick, no annotation: T silently becomes unknown, no error below
const B = lazyLoad(() => Promise.resolve({}), { pick: (m) => ({ default: m.NamedComponent }) });
const b: string = B; // BUG: should also error, but does not
```

[Playground](https://www.typescriptlang.org/play/?ts=5.9.3#code/declare%20function%20lazyLoad%3CT%3E%28%0A%20%20factory%3A%20%28%29%20%3D%3E%20Promise%3Cany%3E%2C%0A%20%20opts%3A%20%7B%20pick%3A%20%28mod%3A%20any%29%20%3D%3E%20%7B%20default%3A%20T%20%7D%20%7D%0A%29%3A%20T%3B%0A%0Adeclare%20const%20NamedComponent%3A%20%28props%3A%20%7B%20foo%3A%20string%20%7D%29%20%3D%3E%20any%3B%0A%0A//%20Extracted%20%2B%20explicitly%20annotated%20pick%3A%20infers%20T%20correctly%0Aconst%20pickFn%20%3D%20%28m%3A%20any%29%3A%20%7B%20default%3A%20typeof%20NamedComponent%20%7D%20%3D%3E%20%28%7B%20default%3A%20m.NamedComponent%20%7D%29%3B%0Aconst%20A%20%3D%20lazyLoad%28%28%29%20%3D%3E%20Promise.resolve%28%7B%7D%29%2C%20%7B%20pick%3A%20pickFn%20%7D%29%3B%0Aconst%20a%3A%20string%20%3D%20A%3B%20//%20correctly%20errors%3A%20%28props%29%20%3D%3E%20any%20is%20not%20string%0A%0A//%20Inline%20pick%2C%20no%20annotation%3A%20T%20silently%20becomes%20unknown%2C%20no%20error%20below%0Aconst%20B%20%3D%20lazyLoad%28%28%29%20%3D%3E%20Promise.resolve%28%7B%7D%29%2C%20%7B%20pick%3A%20%28m%29%20%3D%3E%20%28%7B%20default%3A%20m.NamedComponent%20%7D%29%20%7D%29%3B%0Aconst%20b%3A%20string%20%3D%20B%3B%20//%20BUG%3A%20should%20also%20error%2C%20but%20does%20not%0A)

In our case this means `React.lazy`-based components loaded via a named-export `pick` option silently lose all prop type-checking (`IntrinsicAttributes` errors, or worse, no errors at all where there should be). The workaround (extracting `pick` into a separately-typed `const` at every call site) works but is significant boilerplate for what's otherwise a clean inline pattern — real impact for anyone building this kind of "lazy + adapter" helper.

Just want to add a data point that this is still very much alive and biting real-world code, in case it helps with prioritization.
